### ISSUE TITLE
Email IMAP fetch stops working if mailserver is once unreachable

### ISSUE THREAD

**Issue by hiaz** [zammad/zammad#1861]
### Infos:

* Used Zammad version:  2.3

### Expected behavior:

If I reboot the mailserver e.g. for an update zammad should continue fetching emails via IMAP if the mailserver is up again.

### Actual behavior:

If I reboot the mailserver e.g. for an update zammad stops fetching emails via IMAP even if the mailserver is up again. I have to restart the zammad service.

### Steps to reproduce the behavior:

* Configure an IMAP email channel
* Stop the mailserver
* Drink a cup of coffee or tea (slowly. Just to be sure. Espresso is not suitable)
* (Alternatively - if you are not thirsty) Work on https://github.com/zammad/zammad/issues/223
* Start the mailserver again
* Zammad will not continue to fetch emails

In the case of a mailserver update it is not critical if you know how zammad reacts but if you have a short outtake of the IP connection somewhere between zammad and the mailserver or you outsource the maintenance of the mailserver it may take a longer period until you find out that you do not get any new email tickets.

"Yes I'm sure this is a bug and no feature request or a general question.": This sentence is not mine but somehow true ;).

**Comment by thorsteneckel** [comment]
Hi @hiaz ,

thanks for the great issue report. I did as you say and enjoyed my tea ;)

However, unfortunately I was not able to reproduce this issue. I configured an IMAP account and blocked the connection to it via my firewall for 15 minutes and after that for 30 minutes. After the firewall unblocked the connection Zammad immediately retried the connection. While the connection was not possible I had the error shown in my email configuration view:

<img width="832" alt="bildschirmfoto 2018-03-08 um 10 49 09" src="https://user-images.githubusercontent.com/3873515/37146466-1c8daa62-22c4-11e8-8caa-b4e0ff5e3c8c.png">

And my monitoring listed the error too:
<img width="1052" alt="bildschirmfoto 2018-03-08 um 11 27 16" src="https://user-images.githubusercontent.com/3873515/37146485-2c62ff64-22c4-11e8-8a3c-d130fae4cfb1.png">

Am I doing something wrong here? Can you provide your `log/production.log` file? You can send it to support@zammad.com directly. Please refer to me and this issue. Thanks!

**Comment by MrGeneration** [comment]
I rebootet and updates my E-Mail-Server last weekend (downtime about 15mins) and did not run in this problem, so kinda not able to reproduce too - can test again if wanted.

**Comment by hiaz** [comment]
Hi @thorsteneckel,

thank you for testing and thank you @MrGeneration for verifying. In our situation zammad does not even recognise that the server is down. We updated zammad for many times now - maybe it helps if I delete and recreate the imap account. 

I test this in the next maintenance window and come back with more details.

**Comment by MacLemon** [comment]
I have the same problem on 2.2. When the email server becomes unreachable for Zammad for some time it stops polling the account altogether. 

Workaround:
I have to edit email integrations a re-save the account to make it fetch email again.

Yes, this is reflected in the health URL response.

The scheduler does not seem to retry fetching email on an account it could not reach at least once.

I would have expected this to automatically resume operation when the email server can be reached again.

**Comment by hiaz** [comment]
Tried it again: This time (15min downtime) everything worked like expected.
1. Zammad marked the account with red
2. Zammad started to fetch mails again.
So maybe it depends how long the mailserver is not available. I'll try to get more details if it happens again. Can take some time.

**Comment by thorsteneckel** [comment]
If somebody has the case that this happens and we can execute some statements on the Zammad console to get a better picture on what's currently going on let me know. We are ready.

**Comment by MrGeneration** [comment]
My VPN-Connection died for several hours on 13th April. I happened to notice today that this peaceful silence comes from zammad not fetching my emails and not from no emails arriving.

I did not touch the system - if we can take a look within tomorrow I don't mind not changing anything :) You're invited to take a look ;)

**Comment by thorsteneckel** [comment]
@MrGeneration  - Would love to have a look tomorrow! When does it fit for you? I'm free except 15-16 o'clock. You can reach me directly on my mobile if you want to. Thanks!

**Comment by MrGeneration** [comment]
@thorsteneckel great, I think I'll call ya about 9am or 12am - whatever fits best. :)

**Comment by MrGeneration** [comment]
My Problem has been partly solved by disabling facebook - my zammad instance seemed to be too confused and the scheduler crashed/stopped because of this. After reboot, zammad fetched E-Mails exactly once and then stopped again.

Disabling facebook fixed the problem at my side, at least for now. Maybe the information in my logfile were good enough so the error can be found and fixed.

**Comment by uli-heller** [comment]
I've created #2013 which supposedly is a duplicate of this. For my issue, the imap server was available all the time. I didn't setup anything related to facebook. I restarted the scheduler process and it doesn't help, zammad isn't fetching imap emails. 

When I reconfigure the email channel, I get this:

```
I, [2018-05-10T15:53:08.797272 #2754]  INFO -- : Processing by ChannelsEmailController#verify as JSON
I, [2018-05-10T15:53:08.797434 #2754]  INFO -- :   Parameters: {"inbound"=>{"adapter"=>"imap", "options"=>{"host"=>"xxx", "user"=>"xxx", "password"=>"[FILTERED]", "ssl"=>true, "port"=>"993", "folder"=>"", "keep_on_s
I, [2018-05-10T15:53:13.887752 #2754]  INFO -- : fetching imap (xxx/xxx port=993,ssl=true,folder=INBOX,keep_on_server=false)
I, [2018-05-10T15:53:14.119036 #2754]  INFO -- : verify mode, fetch no emails #80824621106
I, [2018-05-10T15:53:14.159220 #2754]  INFO -- :  - verify email #80824621106 found
I, [2018-05-10T15:53:14.261993 #2754]  INFO -- : Completed 200 OK in 5464ms (Views: 0.3ms | ActiveRecord: 5.0ms)
```

**Comment by uli-heller** [comment]
I just realized the Monitoring screen. It shows like this:

![auswahl_206](https://user-images.githubusercontent.com/1677414/39874244-8fdfe0e0-546d-11e8-97cb-58787c171b97.png)

The three channels are three inbound imap connections. All looks good. What can I do?

**Comment by LittleNo** [comment]
Same here, IMAP import stops after the connection to IMAP-server had been not available. I can reproduce it due to an instable internet connection. Zammad does not start to fetch Emails when the IMAP-server becomes back available. Restarting the zammad-service makes it immediately continue fetching mails.

**Comment by thorsteneckel** [comment]
@LittleNo - that's really interesting because this is exactly what I tried and wasn't able to reproduce. So you can reproduce it every time you want? Would you mind if I send you a debugging script to analyze what's happening under the hood?

**Comment by LittleNo** [comment]
yes let's do that

I have not been watching every single time that issue happened, but when I compare the log of the internet-connection with the moments when I had to restart zammad, it looks like it happens almost every time

**Comment by thorsteneckel** [comment]
Cool, thanks! We'll come back to you in one of our next free time slots.

**Comment by MrGeneration** [comment]
I actually can reproduce this every 2-3 reboots of my OpenVPN-Server-Host.
This reboot takes up  to 5 minutes until the channel is back.

During this period, the network zammad tries to grab the E-Mails from is unreachable (because it's supposed to get it via VPN) and the service tries on reconnecting.

Maybe a hint. I can take a look into the logs the next time it happens.

**Comment by thorsteneckel** [comment]
Fixed by @martini with the commits above. Closing now. Feedback is highly welcome 😍
